File: src/main/java/com/eternalcode/plots/member/MemberRepositoryInMemory.java
package com.eternalcode.plots.member;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

public class MemberRepositoryInMemory implements MemberRepository {

    private final Map<UUID, Set<Member>> membersByPlotUUID = new HashMap<>();

    @Override
    public CompletableFuture<Void> saveMember(Member member) {
        Set<Member> indexedMembers = this.membersByPlotUUID.computeIfAbsent(member.plotId(), key -> new HashSet<>());
        indexedMembers.add(member);

        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletableFuture<Void> removeMember(Member member) {
        Set<Member> indexedMembers = this.membersByPlotUUID.computeIfAbsent(member.plotId(), key -> new HashSet<>());
        indexedMembers.remove(member);

        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletableFuture<Boolean> isMember(UUID plotId, UUID userId) {
        Set<Member> indexedMembers = this.membersByPlotUUID.computeIfAbsent(plotId, key -> new HashSet<>());
        
        return CompletableFuture.completedFuture(indexedMembers.stream().anyMatch(member -> member.userId().equals(userId)));
    }
}

File: src/main/java/com/eternalcode/plots/member/MemberRepositoryInMemory.kt
package com.eternalcode.plots.member

import java.util.*
import java.util.concurrent.CompletableFuture

class MemberRepositoryInMemory : MemberRepository {
    private val membersByPlotUUID: MutableMap<UUID, MutableSet<Member>> = HashMap()

    override fun saveMember(member: Member): CompletableFuture<Void?>? {
        val indexedMembers =
            membersByPlotUUID.computeIfAbsent(member.plotId) { key: UUID? -> HashSet() }
        indexedMembers.add(member)

        return CompletableFuture.completedFuture(null)
    }

    override fun removeMember(member: Member): CompletableFuture<Void?>? {
        val indexedMembers =
            membersByPlotUUID.computeIfAbsent(member.plotId) { key: UUID? -> HashSet() }
        indexedMembers.remove(member)

        return CompletableFuture.completedFuture(null)
    }

    override fun isMember(plotId: UUID, userId: UUID): CompletableFuture<Boolean>? {
        val indexedMembers: Set<Member> =
            membersByPlotUUID.computeIfAbsent(plotId) { key: UUID? -> HashSet() }

        return CompletableFuture.completedFuture(
            indexedMembers.stream().anyMatch { member: Member -> member.userId == userId })
    }
}
